scipy-1.0/ckdtree.tex

```diff
@@ -2,22 +2,20 @@
 periodic boundary conditions was added. A periodic boundary condition is typically 
 used in computer simulations of physical processes.
 
-% TODO: please check / improve literature citations used here
-% what about the time complexity of the kNN implementation?
-In 2015, we enhanced \texttt{cKDTree.query} with a $k$ nearest neighbors search
-parameter. This is an efficient operation\cite{Sproull:1991:RNS:3118219.3118331} 
-because it generates and returns a data structure that may be only a fraction 
-of the size of the full neighbor list. A C++ implementation is provided that releases
-the Python global interpreter lock for the neighbor search process, uses
-a memory pool to allocate and automatically reclaim \texttt{structs}, and
-heapsort to sort the $k$ nearest neighbors distances, even in the case
-of periodic boundary conditions. When $k$ is a non-contiguous list of nearest
-neighbors to account for (i.e., \texttt{[1, 3, 4]}), the intervening distances
-between the nearest neighbor ($k = 1$) and the farthest neighbor requested
-are all interrogated, but only those neighbors that are requested are retained
-in memory. Thus, the maximum heap size for this algorithm is conceptually determined
-as \texttt{np.arange(max(k))}.
+Specifically, in 2015, we enhanced \texttt{cKDTree.query} by reimplementing in C++,
+removing some of the algorithm memory leaks, and allowing 
+release of the GIL so that multiple threads may be used\cite{gh-4374}. In 2013,
+the time complexity of the k nearest neighbor search from \texttt{cKDTree.query}
+was approximately loglinear \cite{knn-jake}, and we've confirmed that this time
+complexity has been preserved in SciPy \texttt{1.0.0} (Figure~\ref{fig:knn-complexity}),
+but with generally improved efficiency (Figure~\ref{fig:asvbench}). The
```

Comment:
Maybe reversing the order of these would flow better?
```
In 2013, the time complexity of the k nearest neighbor search from \texttt{cKDTree.query}
was approximately loglinear \cite{knn-jake}. Since then, we've enhanced \texttt{cKDTree.query} by reimplementing in C++, removing memory leaks, and allowing release of the GIL so that multiple threads may be used\cite{gh-4374}, generally improving performance on any given problem (Figure~\ref{fig:asvbench}) while preserving the asymptotic complexity (Figure~\ref{fig:knn-complexity}).
```
I can't see the figures right now, so I hope that's what was intended.
Re: using multiple threads - for this problem, is it possible for multiple threads on a single to core to improve performance, or should we be more specific e.g. "so that multiple cores can be used simultaneously".

Reply:
Yeah, I've used something similar to your suggestion in the latest set of revisions here now.

I think theoretically releasing the GIL can empower usage of multiple threads on a single physical core that has multiple logical cores for use with hyperthreading, but in practice that may not have been the primary motivation. I haven't changed that specific detail, for now.